### PR TITLE
Staging for __shared and __owned

### PR DESCRIPTION
**Pull request by CodaFi**
This is the bare minimum to get the frontend to accept `__shared` and `__owned`.  `@guarantee`d semantics and `__owned`<=>`__shared` conversions are coming next.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: include/swift/Demangling/DemangleNodes.def [hunk 1]
```diff
@@ -135,6 +135,7 @@ NODE(QualifiedArchetype)
 NODE(ReabstractionThunk)
 NODE(ReabstractionThunkHelper)
 NODE(ReturnType)
+NODE(__Shared)
```
Comment (eeckstein) [inline include/swift/Demangling/DemangleNodes.def]: No need to add underscores

File: docs/ABI.rst [hunk 2]
```diff
@@ -1039,7 +1039,7 @@ Types
   type-list ::= list-type '_' list-type*     // list of types
   type-list ::= empty-list
 
-  list-type ::= type identifier? 'z'? 'd'?   // type with optional label, inout convention and variadic specifier
+  list-type ::= type identifier? 'z'? 'h'? 'd'?   // type with optional label, inout convention, shared convention, and variadic specifier
```
Comment (eeckstein) [inline docs/ABI.rst]: Basically this is ok. But we have to be careful about filling up the operator namespace.
If we expect that shared is only rarely used, it's not worth "wasting" a whole character 'h'.
Instead we could do something like using 'dv' as a variadic marker (which also does not occur too often) and use 'ds' for the shared convention.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Review by eeckstein — COMMENTED** [review]
Overall the mangling changes lgtm.

File: docs/ABI.rst (hunk 2 above)
Comment (rjmccall) [inline docs/ABI.rst]: I expect shared to be common enough to warrant abbreviation.  It'll be used heavily in the standard library if nothing else.

File: include/swift/Demangling/DemangleNodes.def (hunk 1 above)
Comment (CodaFi) [inline include/swift/Demangling/DemangleNodes.def]: I fully expect this name to change.  Even given that, should I remove them?

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: docs/ABI.rst (hunk 2 above)
Comment (CodaFi) [inline docs/ABI.rst]: Spoke to @eeckstein in person.  We agreed to try the single-letter mangling and revisit a two-char mangling at a later date.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
Last one, I swear 

@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
⛵️

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**